util/wasmstorerebuilder/wasmstorerebuilder.go

```diff
@@ -0,0 +1,106 @@
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/offchainlabs/nitro/arbos/arbosState"
+)
+
+var RebuildingPositionKey []byte = []byte("_rebuildingPosition")             // contains the codehash upto where rebuilding of wasm store was last completed
+var RebuildingStartBlockTimeKey []byte = []byte("_rebuildingStartBlockTime") // contains the block time when rebuilding of wasm store first began
+var RebuildingDone common.Hash = common.BytesToHash([]byte("_done"))         // indicates that the rebuilding is done, if RebuildingPositionKey holds this value it implies rebuilding was completed
+
+func GetRebuildingParam[T any](wasmStore ethdb.KeyValueStore, key []byte) (T, error) {
+	var empty T
+	posBytes, err := wasmStore.Get(key)
+	if err != nil {
+		return empty, err
+	}
+	var val T
+	err = rlp.DecodeBytes(posBytes, &val)
+	if err != nil {
+		return empty, fmt.Errorf("invalid value stored in rebuilding key or error decoding rebuildingBytes: %w", err)
+	}
+	return val, nil
+}
+
+func SetRebuildingParam[T any](wasmStore ethdb.KeyValueStore, key []byte, val T) error {
+	valBytes, err := rlp.EncodeToBytes(val)
+	if err != nil {
+		return err
+	}
+	err = wasmStore.Put(key, valBytes)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// RebuildWasmStore function runs a loop looking at every codehash in diskDb, checking if its an activated stylus contract and
+// saving it to wasm store if it doesnt already exists. When errored it logs them and silently returns
+//
+// It stores the status of rebuilding to wasm store by updating the codehash (of the latest sucessfully checked contract) in
+// RebuildingPositionKey every 50 checks.
+//
+// It also stores a special value that is only set once when rebuilding commenced in RebuildingStartBlockTimeKey as the block
+// time of the latest block when rebuilding was first called, this is used to avoid recomputing of assembly and module of
+// contracts that were created after rebuilding commenced since they would anyway already be added during sync.
+func RebuildWasmStore(ctx context.Context, wasmStore ethdb.KeyValueStore, l2Blockchain *core.BlockChain, key common.Hash, rebuildingStartBlockTime uint64) {
+	latestHeader := l2Blockchain.CurrentBlock()
+	latestState, err := l2Blockchain.StateAt(latestHeader.Root)
+	if err != nil {
+		log.Error("error getting state at latest block, aborting rebuilding", "err", err)
+		return
+	}
+	diskDb := latestState.Database().DiskDB()
+	arbState, err := arbosState.OpenSystemArbosState(latestState, nil, true)
+	if err != nil {
+		log.Error("error getting arbos state, aborting rebuilding", "err", err)
+		return
+	}
+	programs := arbState.Programs()
+	iter := diskDb.NewIterator(rawdb.CodePrefix, key[:])
+	for count := 1; iter.Next(); count++ {
+		// If outer context is cancelled we should terminate rebuilding. We probably wont be able to save codehash to wasm store (it might be closed)
```

Comment:
when we do a parallel RebuildWasm version - it should be done in executionEngine, and then it should have a context from ExecutionEngine that will be stopped before closing the database if done orderly..
Try writing that last value, Worst case - it won't work